Loading Option added to OptionsSelector

Santhosh-Sellavel opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
Reduced SearchModal transition duration by using stack navigation options. 

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/4723
--->
$ https://github.com/Expensify/App/issues/4822

### Tests & QA Steps
1. Login into new Expensify app in Desktop/Web
2. Use shortcut `CMD + K` or `CTRL + K` to enable search
3. Search Modal will load faster now. 
4. So able to capture search inputs now bit quicker.

### Tested On
- [X] Web
- [X] Desktop
- [ ] iOS
- [ ] Android
- [ ] Desktop

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Desktop
https://user-images.githubusercontent.com/85645967/136999403-a44f84d5-91f2-4af2-86bf-116253525e21.mov

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

Santhosh-Sellavel committed:
Loading Option added to OptionsSelector

Santhosh-Sellavel commented on src/components/OptionsSelector.js:
```diff
@@ -94,7 +98,15 @@ class OptionsSelector extends Component {
 
     componentDidMount() {
-        this.textInput.focus();
+        if (!this.props.isLoading) {
+            this.textInput.focus();
+        }
```
@Luke9389  Remove the check focus will be set to text input immediately, but the transition animation will not happen view will be presented immediately, receive input from the user immediately

Luke9389 replied:
Sorry, I don't exactly understand why this change was made. Don't we want the focus to happen immediately?

Santhosh-Sellavel replied:
Since we use transition animation, setting `focus` to `textInput` immediately will stop the transition abruptly.

Luke9389 commented:
Hey @Santhosh-Sellavel, I'll be reviewing this tomorrow. Thanks for your patience :)

Santhosh-Sellavel committed:
Reduced animation duration so it will be responsive quickly

Santhosh-Sellavel commented:
@Luke9389 Test this out, 
Reduced animation duration so it will be responsive quickly -> https://github.com/Expensify/App/pull/4903/commits/a649d45cb8046e71ea791d42b82276acb5b8d019

Note:
We can refactor this as you suggest.

Santhosh-Sellavel commented:
@Luke9389 Any update on this?

Luke9389 commented on src/components/OptionsSelector.js:
```diff
@@ -203,21 +215,26 @@ class OptionsSelector extends Component {
+                            forceTextUnreadStyle={this.props.forceTextUnreadStyle}
+                            showTitleTooltip={this.props.showTitleTooltip}
+                        />
+                    ) : <FullScreenLoadingIndicator isVisible={this.props.isLoading} />}
```
Why do we have `isVisible={this.props.isLoading}`? We don't render the `<FullScreenLoadingIndicator/>` unless we are loading right?

Santhosh-Sellavel replied:
Sorry my bad, it's unnecessary, will remove it.

Luke9389 replied:
Please do that then.

Luke9389 commented on src/libs/Navigation/AppNavigator/ModalStackNavigators.js:
```diff
@@ -113,9 +114,31 @@ const ReportParticipantsModalStackNavigator = createModalStackNavigator([
 
+// https://reactnavigation.org/docs/stack-navigator/#options
+// We can refactor and move this somewhere else
+const searchNavigationOptions = {
+    transitionSpec: {
+        open: {
+            animation: 'timing',
+            config: {
+                duration: 200,
+                delay: 0,
+            },
+        },
+        close: {
+            animation: 'timing',
+            config: {
+                duration: 300,
+                delay: 0,
+            },
+        },
+    },
+};
```
@parasharrajat I noticed some discussion in the slack thread you posted that mentions the smoothness of the in and out animation. Is this what you had in mind? 

I notice the `duration`s are different. Why?

parasharrajat replied:
Sorry, This change is not what I meant. Also, I don't think this change is acceptable. Changing animation for one of the pages is not good. Page animation should be the same for all pages in the App. 

Calling Gandalf @shawnborton. 

Extraneous, This PR is still in draft and so let the OP prepare it and mark is ready for review. IMO, we should not review a draft PR.

Santhosh-Sellavel replied:
> I notice the `durations` are different.
I don't know I just tested with different timings. We should unify this definitely.

Santhosh-Sellavel replied:
> Sorry, This change is not what I meant. Also, I don't think this change is acceptable.

I will add a follow-up comment shortly in the PR, explaining why. Please wait!

> Changing animation for one of the pages is not good. Page animation should be the same for all pages in the App.

I agree on this, I was waiting for @Luke9389 to see my changes. I was planning to a discuss on this once he is back. I will address more in a follow-up comment.

> Calling Gandalf @shawnborton.
I was planning too, thanks for this.

> Extraneous, This PR is still in draft and so let the OP prepare it and mark is ready for review. IMO, we should not review a draft PR.

I am not sure about this, but it's okay to check because see if there is any progress or for any clarifications.

Luke9389 replied:
> Changing animation for one of the pages is not good. Page animation should be the same for all pages in the App.
> 
> > I agree on this, I was waiting for @Luke9389 to see my changes. I was planning to a discuss on this once he is back. I will address more in a follow-up comment.

Ok, so we all agree they should not be set here. Where are animation durations being set currently?

Santhosh-Sellavel replied:
Just below the declaration. 
For the search screen, transition animation is part of the navigation options passed it as variable options shown below,
https://github.com/Expensify/App/blob/a649d45cb8046e71ea791d42b82276acb5b8d019/src/libs/Navigation/AppNavigator/ModalStackNavigators.js#L138-L142

used In `createModalStackNavigator` method in the line below
https://github.com/Expensify/App/blob/a649d45cb8046e71ea791d42b82276acb5b8d019/src/libs/Navigation/AppNavigator/ModalStackNavigators.js#L57

This is how animation is set.

Santhosh-Sellavel replied:
> Ok, so we all agree they should not be set here. 

Yes, not like how I did here.

Should move these [lines](https://github.com/Expensify/App/blob/a649d45cb8046e71ea791d42b82276acb5b8d019/src/libs/Navigation/AppNavigator/ModalStackNavigators.js#L117-L136) to a separate file, 
1. Refactor it as a function to make it reusable for other screens providing customize options based on each screen & use it for those screens alone. 

2. Just create a CONST variable with a common name use it here or anywhere else if there is no screen-specific customization need case.

Luke9389 replied:
Rather than just suggesting these changes in a comment, please do them so we can see the code changes.

Luke9389 commented on src/pages/SearchPage.js:
```diff
@@ -170,31 +170,25 @@ class SearchPage extends Component {
-                        <KeyboardSpacer />
-                    </>
-                )}
+            <ScreenWrapper onTransitionEnd={() => this.setState({isLoading: false})}>
```
Does this assume loading will be done by the end of the transition?

Santhosh-Sellavel replied:
As of now (even without this PR), we show loading only for the transition period.  We could refactor it when we plan to use it for other actions like asynchronous data fetch, etc.

Luke9389 replied:
Ok, so we are showing loading **because** the transition is happening?

Santhosh-Sellavel replied:
Yes @Luke9389 , see here.
In our current app, the following components where we showing loading for transition alone
 1: Search Page
Search Component,  until `didScreenTransitionEnd` we show `FullScreenLoadingIndicator` & hide the OptionsSelector. https://github.com/Expensify/App/blob/193ebc91dc1ecfd77fd5cbf0ee313afb32fc5dc7/src/pages/SearchPage.js#L174-L195

 2: New chat page
https://github.com/Expensify/App/blob/193ebc91dc1ecfd77fd5cbf0ee313afb32fc5dc7/src/pages/NewChatPage.js#L132-L144

 3: New group page
https://github.com/Expensify/App/blob/193ebc91dc1ecfd77fd5cbf0ee313afb32fc5dc7/src/pages/NewGroupPage.js#L195-L209

4: IOU Modal
https://github.com/Expensify/App/blob/193ebc91dc1ecfd77fd5cbf0ee313afb32fc5dc7/src/pages/iou/IOUModal.js#L322-L327

Luke9389 commented:
Hey @Santhosh-Sellavel, thanks for waiting on this. I've been out of the office recently and busy with other stuff, so I hadn't had a chance to loop back around to this. I've left a few questions that I don't know the answer to.

Santhosh-Sellavel commented:
> Hey @Santhosh-Sellavel, thanks for waiting on this. I've been out of the office recently and busy with other stuff, so I hadn't had a chance to loop back around to this. I've left a few questions that I don't know the answer to.

I will answer all shortly will let you once done, we can discuss if possible. @Luke9389

Santhosh-Sellavel committed:
Merge branch 'main' into Search_Quicker_Response

Santhosh-Sellavel commented:
## Explanations 
The requirement was to render the search component first to receive inputs first & load the contact list bit later, to address the issue after the user presses `CMD + K`, and **immediately starts typing input we need to capture.** 

I had initially proposed to remove the animation it will solve the issue. We decided to keep animation too.  

In commit https://github.com/Expensify/App/pull/4903/commits/6b02bca682da168ecf884eb5cb1768223f91dc19

-  I added loading property to render list later or show loading indicator while transitioning. 
-  Based on the changes component Search will render first & the list will render after the transition ends.
-  But user input is not captured yet again until the transition end. 
- To capture input we need to set focus to text input. But * if we set the focus immediately it will end the transition** immediately as I said here -> https://github.com/Expensify/App/pull/4903#discussion_r698354320

We need animation & also should capture input soon enough. 
To address I added this commit https://github.com/Expensify/App/pull/4903/commits/a649d45cb8046e71ea791d42b82276acb5b8d019

- Now we have control to speed up and slow down the loading & dismiss animation, **so we can get input as soon as possible**. 
- `200ms` is speed enough I think, I typed as fast as I can I didn’t notice any missing characters.
- We need to refactor this to make it usable for other components too.
- We can use this commit alone (To receive input fast enough & https://github.com/Expensify/App/pull/4903/commits/6b02bca682da168ecf884eb5cb1768223f91dc19 to render search before transition)

Also, this change should be done only when we use `CMD + K` which is a command. Just a suggestion maybe we can add a new root for this shortcut to differentiate timing and load fast enough.  If there are any other without creating a new root to identify this from shortcut & or from search action let me know.

Note: Across other apps commonly, I see there is no animation at all or fast animations when we use shortcuts to perform a search.

I see I addressed everything I can. If there are any other ways to address this issue or my approach totally is wrong. Let's make this issue open for proposals again.

cc: @Luke9389  @parasharrajat

Santhosh-Sellavel commented:
@Luke9389  Please post your thoughts I updated the comments.

kadiealexander commented:
Hi @Santhosh-Sellavel, I've placed this issue on hold as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1631754482374800), as we are prioritising issues related to a feature release scheduled for 9/24. As an apology for the delay, we will add a $100 bonus as a thank you for waiting.

Luke9389 commented:
Hey @Santhosh-Sellavel, any update on this. You can go ahead and take this off draft to begin the reviewing rounds.

Santhosh-Sellavel commented:
@Luke9389 
Have a look at this comment -> https://github.com/Expensify/App/pull/4903#issuecomment-916606975
Waiting for your thoughts inputs to proceed further

Santhosh-Sellavel commented:
Thanks @Luke9389 

Will implement the changes & mark PR Ready for review.

Santhosh-Sellavel committed:
Merge branch 'main' into Search_Quicker_Response

Santhosh-Sellavel committed:
Refactored the transitionspec into a new file

Santhosh-Sellavel committed:
Unnecessary prop removed

Santhosh-Sellavel committed:
Merge branch 'main' into Search_Quicker_Response

Santhosh-Sellavel commented:
@Luke9389,
I've had done the refactoring part here. Marking it ready for review.

Two things :
1. Do we need to maintain same transition duration in mobile also?
I'm confused here, so currently duration is same across all screens. 
If not needed let me I'll make necessary changes along with any review changes.

2. If we still feel this it is not quick enough to capture inputs?
IMO We should disable the animation for search view. 
I've seen few apps on using search shortcut, Search view or bar appears almost instantly. 

Thanks

cc: @parasharrajat

marcaaron commented:
Can the `[HOLD]` be removed now on this? What's the status?

Santhosh-Sellavel commented:
> > Do we need to maintain same transition duration in mobile also?
> 
> I would assume the answer here is yes.

Then this PR is ready for review. 

> > If we still feel this it is not quick enough to capture inputs?
> > IMO We should disable the animation for search view.

> I'm not sure on this one. We could have a discussion about removing the search view animation in the `#expensify-open-source` Slack channel. Would you be willing to post there regarding this?

Sure I'll start a discussion if needed. But first if possible test the PR changes and let me know if there is any lag?

cc: @Luke9389 @marcaaron

Santhosh-Sellavel committed:
Merge pull request #3 from Santhosh-Sellavel/main

Pull from main
